docs/source/project_documentation/dev_guide/aac_constraints.md

```diff
@@ -0,0 +1,395 @@
+---
+layout: default
+parent: Developer's Guide to AaC
+title: AaC Constraint Checks
+nav_order: 3
+has_children: false
+---
+
+# What is Constraint Checking in the AaC Language?
+Because the AaC DSL is leveraging plain-text YAML as the underpinning of the DSL, there is little to no functionality to guide users in the correctness of their YAML AaC structures. AaC has implemented a self-checking language feature so that users can reference which rules are applied to which AaC DSL components, and so that users can define constraints for their own user-defined structures. To this end, AaC employs a plugin-based constraint system where plugins provide Python-based constraint implementations that can be referenced and applied to definitions in the AaC DSL.
+
+Constraint rules in AaC are defined with the `context_constraint`, `schema_constraint`, and `primitive_constraint` definitions, which are required to have a corresponding implementation, via an AaC plugin. This enables AaC's self-checking mechanism even though YAML is just a markup language.
+
+## Checking an AaC File
+The overall constraint mechanism follows this flow:
+1. A definition is identified as needing to be checked (i.e. you run `aac check <AaC_file_path>`)
+2. Certain "core" constraints must be automatically checked while parsing and loading the AaC definition.
+  a. This includes items such as required fields being present, fields being recognized based on their defining schema, etc.
```

Comment:
I updated the text to remove the word "constraints" from this entry to try and avoid confusion.  While I agree it may be ideal to handle these items as "proper constraints" defined via plugins, it just doesn't seem reasonable.  It's a catch 22 situation where you need to be able to make sense of the data before you can hand it off to a "proper constraint" for further checking.  If there's a way to do this without dramatically increasing complexity, I'm all for putting it in later.